### ISSUE TITLE
Energy Conservation of MTSIntegrator

### ISSUE THREAD

**Issue by kyleabeauchamp** [openmm/openmm#1616]
Is there a canonical way to test the energy conservation performance of our integrators?  I'm finding reduced conservation as compared to Verlet, as measured by the below script.

```
import simtk.openmm.app as app
import simtk.openmm as mm
from simtk import unit as u
from openmmtools import testsystems

f = lambda x: (x.getPotentialEnergy(), x.getKineticEnergy(), x.getKineticEnergy() + x.getPotentialEnergy())

temperature = 300. * u.kelvin
testsystem = testsystems.LennardJonesFluid(nparticles=2048)
timestep = 0.1 * u.femtoseconds

integrators = [mm.VerletIntegrator(timestep), mm.MTSIntegrator(timestep, groups=[(0, 1)])]

for integrator in integrators:
    print(integrator)

    simulation = app.Simulation(testsystem.topology, testsystem.system, integrator)
    simulation.context.setPositions(testsystem.positions)
    simulation.context.setVelocitiesToTemperature(temperature)

    integrator.step(1)  # Do a single step first to apply constraints before we run tests.

    state1a = simulation.context.getState(getPositions=True, getParameters=True, getEnergy=True, getVelocities=True)
    old = f(state1a)
    print("PE, KE, TOTAL")
    print(old)
    integrator.step(1)
    state1b = simulation.context.getState(getPositions=True, getParameters=True, getEnergy=True, getVelocities=True)
    new = f(state1b)
    print(new)
    print("delta = %s" % (new[-1] - old[-1]))


<simtk.openmm.openmm.VerletIntegrator; proxy of <Swig Object of type 'OpenMM::VerletIntegrator *' at 0x7fbcdd422f60> >
PE, KE, TOTAL
(Quantity(value=-220.9337554729007, unit=kilojoule/mole), Quantity(value=7704.49522032775, unit=kilojoule/mole), Quantity(value=7483.561464854849, unit=kilojoule/mole))
(Quantity(value=-220.95013973622264, unit=kilojoule/mole), Quantity(value=7704.512487734668, unit=kilojoule/mole), Quantity(value=7483.562347998445, unit=kilojoule/mole))
delta = 0.0008831435961837997 kJ/mol
<simtk.openmm.mtsintegrator.MTSIntegrator; proxy of <Swig Object of type 'OpenMM::CustomIntegrator *' at 0x7fbcb9059fc0> >
PE, KE, TOTAL
(Quantity(value=-220.8974362141165, unit=kilojoule/mole), Quantity(value=7677.541015625, unit=kilojoule/mole), Quantity(value=7456.643579410884, unit=kilojoule/mole))
(Quantity(value=-220.8767622218149, unit=kilojoule/mole), Quantity(value=7677.5068359375, unit=kilojoule/mole), Quantity(value=7456.630073715685, unit=kilojoule/mole))
delta = -0.013505695198546164 kJ/mol


```

**Comment by jchodera** [comment]
Two observations about your test:
1. The total energy from step to step can vary greatly, so I'd accumulate a hundred samples or so and look at the standard deviation of the total energy as a more reasonable metric of total energy fluctuations.
2. The `MTSIntegrator` uses velocity Verlet (where position and velocity are in sync), while the `VerletIntegrator` uses the leapfrog Verlet (where position and velocity are out of sync by half a timestep). I believe the `VerletIntegrator` uses a different kinetic energy computation scheme than `MTSIntegrator` because of this, which may affect your comparison.

Our modern understanding of thermostatted integrators that are based on symplectic cores is that they sample from a shadow Hamiltonian and generate a perturbed equilibrium distribution that differs from the target distribution in a timestep-dependent manner. It's not possible to compute the shadow Hamiltonian, though in some cases it's possible to approximate it. An alternative is to use nonequilibrium stat mech to measure how far out of equilibrium the system is driven, which we did [in this paper](http://dx.doi.org/10.1103/PhysRevX.3.011007). @maxentile in our group is working on using these ideas to benchmark different thermostatted integration schemes against each other to see what maximum timestep would lead to the same total nonequilibrium free energy difference.

@peastman has a draft manuscript on appropriate ways to monitor energy drift that may also be relevant in terms of measuring energy conservation.

**Comment by kyleabeauchamp** [comment]
I have a theory on why this happens.  Look at line 104 here: https://github.com/pandegroup/openmm/blob/f7935fef719ed0cd8292afef676c39d881b6a960/wrappers/python/simtk/openmm/mtsintegrator.py#L104

Notice how the velocity is only preserved via the history (x1 -x ) / dt?  I wonder if that calculation involves a loss of precision as compared to the "usual" way of doing things (e.g. when updated incrementally `v = v + delta v`)

**Comment by jchodera** [comment]
> Notice how the velocity is only preserved via the history (x1 -x ) / dt? I wonder if that calculation involves a loss of precision as compared to the "usual" way of doing things (e.g. when updated incrementally v = v + delta v)

That's a great point. The current recommended way to do this is via the [velocity Verlet scheme](https://github.com/choderalab/openmmtools/blob/master/openmmtools/integrators.py#L224-L230):
```
        self.addComputePerDof("v", "v+0.5*dt*f/m")
        self.addComputePerDof("x", "x+dt*v")
        self.addComputePerDof("x1", "x")
        self.addConstrainPositions()
        self.addComputePerDof("v", "v+0.5*dt*f/m+(x-x1)/dt")
        self.addConstrainVelocities()
```

**Comment by kyleabeauchamp** [comment]
I can confirm that after repeating my experiment several times, the magnitude of the MTS drift is always > 100X greater than the Verlet drift.

**Comment by kyleabeauchamp** [comment]
Perhaps we can re-formulate the MTS implementation to use VV and get better conservation?   Is that possible?  I can't recall offhand without digging through the literature.

**Comment by jchodera** [comment]
Apologies for being unclear before: `MTSIntegrator` *does* use velocity Verlet, but it has that precision issue you noted. PR fix coming in a moment.

**Comment by jchodera** [comment]
Oddly enough, there aren't any tests for `MTSIntegrator`. I wonder how it got in there in the first place...

**Comment by kyleabeauchamp** [comment]
After your patch, I see a much smaller drift:

```
<simtk.openmm.openmm.VerletIntegrator; proxy of <Swig Object of type 'OpenMM::VerletIntegrator *' at 0x7f47ab104b10> >
PE, KE, TOTAL
(Quantity(value=-221.32956658217887, unit=kilojoule/mole), Quantity(value=7656.888677824289, unit=kilojoule/mole), Quantity(value=7435.55911124211, unit=kilojoule/mole))
(Quantity(value=-221.97040727582225, unit=kilojoule/mole), Quantity(value=7657.528302460909, unit=kilojoule/mole), Quantity(value=7435.557895185087, unit=kilojoule/mole))
delta = -0.001216057023157191 kJ/mol
<simtk.openmm.mtsintegrator.MTSIntegrator; proxy of <Swig Object of type 'OpenMM::CustomIntegrator *' at 0x7f478b53bcf0> >
PE, KE, TOTAL
(Quantity(value=-219.64110596205376, unit=kilojoule/mole), Quantity(value=7634.296875, unit=kilojoule/mole), Quantity(value=7414.655769037946, unit=kilojoule/mole))
(Quantity(value=-218.5692423799208, unit=kilojoule/mole), Quantity(value=7633.22802734375, unit=kilojoule/mole), Quantity(value=7414.658784963829, unit=kilojoule/mole))
delta = 0.0030159258831190527 kJ/mol
```

**Comment by kyleabeauchamp** [comment]
To be sure, my test was a necessary but not sufficient condition...

**Comment by peastman** [comment]
I'm glad you found a way to improve this.  I should explain, though, that the root cause is completely different from what you think it is: in single precision mode, CustomIntegrator is always less accurate than the standard integrators.  That's because the standard ones use a little bit of double precision even when running in single precision mode.  See https://github.com/pandegroup/openmm/blob/master/platforms/cuda/src/kernels/verlet.cu#L41.  Storing just that one value in double precision turns out to have a big effect on energy drift, so we always do that if at all possible.  But there's no way to tell a CustomIntegrator to use double precision for just one variable, so it's better to use mixed precision mode.

**Comment by jchodera** [comment]
Isn't @kyleabeauchamp running in mixed-precision mode by default? Or does `CustomIntegrator` still use single-precision even in mixed-precision mode?

**Comment by peastman** [comment]
The default mode is single, not mixed.  If you use mixed precision, all integration is done in double precision.  With CustomIntegrator, that's usually a good idea.  With standard integrators, single precision mode is usually good enough, but part of the reason for that is that they still use double precision for that one variable in that one kernel.

**Comment by jchodera** [comment]
@kyleabeauchamp: Some additional fixes went into the `CustomIntegrator` in #1694, and in both #1717 and #1715 may impact your energy conservation. Might be good to test this again and see if we can close this.
